elastic-analytics/OpenSearch-Dashboards
# [POC] init impl to add datasource service/client lifecycle

### Description
- Add dataSourceClient
  - a map `dataSourceClientsPool`
  - a flag that maps to global feature flag in `osd.yml`
  - `asDataSource()` to find/create client, then return to caller
- Add `opensearch_data_service`
  - depends on savedObject service to init a `savedObjectClient`
  - init `DataSourceClient` instance
- Register `opensearch_data_service` to `CoreSetup`, `CoreStart`, `CoreRouteHandlerContext`
-  Refactor `search_source` to add `dataSourceId` field, and add it to `ISearchOptions` to provide to `search()` interface
- Refactor `opensearchagg.ts`, to attach data source id if exists
- Refactor `opensearch_seearch_strategy` to select between default cluster client and data source client

### TODO
- See comment for more. The critical TODOs for the end-to-end use case to run are:
  - Align with the data modeling of `index pattern` and `dataSource`, pending PR from @kristenTian 
 
### Test
Only manual test for now
Steps:
1. add ecommerce sample data
2. delete its index
3. create a valid datasource and credential
4. Manually attach `dataSourceId` to ecommerce index pattern by something like:
  ```
POST /.kibana/_update/index-pattern:ff959d40-b880-11e8-a6d9-e546fe2bba5f
{
  "script": "ctx._source['index-pattern'].dataSourceId = '794b5590-088b-11ed-8694-134ac2453186'"
}
  ```
5. Now you should be able to see a ecommerce visualization with the data from data source
6. Discover/TSVB/Vega/Timelion is not working for now, I'll post other PRs to address those.

### Issues Resolved
TBA
 
### Check List
